File: client/src/main/java/ch/ergon/modern/java/client/shape/ShapeAlgorithms.java
package ch.ergon.modern.java.client.shape;

import ch.ergon.modern.java.api.shape.*;

public class ShapeAlgorithms {

    private final ShapeFactory shapeFactory = ShapeFactory.getInstance();

    public Shape copy(Shape original) {
        return switch (original) {
            case Circle c -> shapeFactory.createCircle(c.getRadius());
            case Square s -> shapeFactory.createSquare(s.getSide());
            case Rectangle r -> shapeFactory.createRectangle(r.getWidth(), r.getHeight());
        };
    }

    public Shape simplify(Shape original) {
        return switch (original) {
            case Rectangle r && r.getWidth() == r.getHeight() -> shapeFactory.createSquare(r.getWidth());
            default -> original;
        };
    }
}

File: client/src/main/kotlin/ch/ergon/modern/java/client/shape/kotlin/ShapeAlgorithms.kt
package ch.ergon.modern.java.client.shape.kotlin

import ch.ergon.modern.java.api.shape.kotlin.*

private val shapeFactory = ShapeFactory.instance

fun Shape.copy(): Shape = when (this) {
    is Circle -> shapeFactory.createCircle(radius)
    is Square -> shapeFactory.createSquare(side)
    is Rectangle -> shapeFactory.createRectangle(width, height)
}

fun Shape.simplify(): Shape = when (this) {
    // Doesn't seem possible to add further conditions to a when in Kotlin
    is Rectangle -> if (width == height) shapeFactory.createSquare(width) else this
    else -> this
}